pkg/pillar/cmd/zedmanager/updatestatus.go

```diff
@@ -421,6 +421,23 @@ func doActivate(ctx *zedmanagerContext, uuidStr string,
 	log.Functionf("doActivate for %s", uuidStr)
 	changed := false
 
+	// Do we try to activate an application earlier than it's configured to start?
+	if config.ShouldDelay && time.Now().Before(config.StartTime) && !status.DelayHandled {
+		// If we try to activate it for the first time - mark is with the corresponding state
+		if status.State != types.START_DELAYED && !status.ActivateInprogress && !status.Activated {
+			status.State = types.START_DELAYED
+			return true
+		}
+		return changed
+	}
+
+	// If we activate a VM with a delay without waiting the start time as we come here too late, mark the delay
+	// as handled here
+	if !status.DelayHandled {
+		changed = true
+	}
+	status.DelayHandled = true
+
```

Comment:
This code should be moved after the network setup i.e., just before the MaybeAddDomainConfig() call below.

Reply:
Is there a reason for that? I'm not against it, I'm just trying to understand the logic.

Reply:
Makes sense to set up the network and reserve memory etc (and getting any errors from those) before applying the delay.

Reply:
Ok, get it, thanks!

Reply:
I just remember my discussion with @giggsoff, and the argument for putting this code in the beginning of the function.
If we move this part later in the doActivate() function, it can be a situation, where we checked for memory, delayed the start, and then started to create new VMs, or do whatever else memory-consuming. The problem here is that after we check the memory, we do not really reserve it. So, it can be dangerous to split these two moments in time: checking for memory availability and the real VM creation.

Reply:
In case of having another application we want to run before delayed one we should not go into memory checking. For example we can have another application controlled by local profile manager that will start and stop in meantime and will consume memory. If it is not expected behaviour, we can move delay check after memory/networks.